### PR TITLE
Generalize repository decoration

### PR DESCRIPTION
**Pull request by DaveCTurner**
We would like to add features to snapshots by decorating existing blob store
repositories, adding encryption, searchability etc, similarly to today's
source-only snapshots. Today's interface does not allow us to combine these
features together in interesting ways.

This commit allows plugins to optionally consume some of the repository
settings and use them to configure a wrapper around the basic blob store
repository.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 1]
```diff
@@ -418,9 +426,32 @@ private Repository createRepository(RepositoryMetaData repositoryMetaData, Map<S
...
+        if (decorators.size() > 1) {
+            throw new RepositoryException(repositoryMetaData.name(),
+                "multiple plugins attempted to decorate repository but this is forbidden: " + decorators);
+        }
+        final RepositoryMetaData repositoryMetaDataAfterDecoration;
+        final Function<Repository, Repository> repositoryWrapper;
+        if (decorators.size() == 1) {
+            Settings.Builder settings = Settings.builder().put(repositoryMetaData.settings());
+            for (final Setting consumedSetting : decorators.get(0).getConsumedSettings()) {
+                settings.remove(consumedSetting.getKey());
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: I guess technically we could do without the setting removal for now, we're not doing any setting validation anyway? Seems like currently it just makes things more complicated to no benefit?

**Review by original-brownbear — COMMENTED** [review]
Thanks David at least from my perspective this is exactly what we discussed last week :)
Just wondering about the settings removal (if we didn't have that we could also simplify away the `getConsumedSettings` for now ... seems to me this was only an idea if we wanted to chain decorators but if we only allow a single one per repo it seems unnecessary?)

**Comment by DaveCTurner** [comment]
Yeah you're right, we can leave the settings alone for now since nobody cares about extra settings; consider that bit to be a demonstration that we can add better validation of the settings in future without breaking this approach.

One thing that isn't obvious here is that if we want to compose repository wrappers from different modules then we'll need an über-module to access them all at the same time. I understood that `xpack/core` was such an über-module but it isn't, it's something on which the rest of xpack can depend (an unter-module?).

**Comment by original-brownbear** [comment]
> One thing that isn't obvious here is that if we want to compose repository wrappers from different modules then we'll need an über-module to access them all at the same time. I understood that xpack/core was such an über-module but it isn't, it's something on which the rest of xpack can depend (an unter-module?).

I guess the question is, do we need that (for now?) or can we just live with searchable snapshots either living in the x-pack core codebase or be a dependency of x-pack core and have things work like they currently do for source only snapshots but add the searchable aspect to the source_only decorator logic?

**Comment by DaveCTurner** [comment]
Yes, that is the question :)

IMO yes I think we should do that now, or else we'll never do it. I don't know exactly what it entails, but if it's really painful then that sounds like a bug in its own right.

**Comment by original-brownbear** [comment]
> IMO yes I think we should do that now, or else we'll never do it. I don't know exactly what it entails, but if it's really painful then that sounds like a bug in its own right.

I think I disagree here. We will only have a limited number of x-pack related repository decorators and those will have to be applied to a repo in a fixed order (encryption first etc.). So if we use the current approach, we'd have to build another interface for things that a module can export and will chain to a repository decorator, including some prioritization mechanism. 
I feel like the fact that we'd need the prioritization makes this so complicated to generalize that it's not really worth it. Who else is going to use it and would we really want to make that kind of thing a stable API?

**Comment by DaveCTurner** [comment]
I wasn't meaning anything so general. We just need a module that can depend on `x-pack/searchable-snapshots`, `x-pack/source-only-snapshots` and `x-pack/encrypted-snapshots` and which implements the wrapping logic directly and explicitly. At the moment IIRC we need things from `x-pack/core` so we can't be a dependency of `x-pack/core`, nor can we have each snapshot plugin as a separate thing with `x-pack/core` as a dependency since they can't know about each other, so we have to put everything in `x-pack/core` itself and that concerns me.

**Comment by original-brownbear** [comment]
> so we have to put everything in x-pack/core itself and that concerns me.

It's not great, but currently source only snapshots are in core as well. So any solution along the lines of creating a separate module that depends on all the snapshot wrappers also entails refactoring source only snapshots out of x-pack/core. 
That seems like a separate effort from what we're doing here. It's not really functionally relevant either, more a matter of moving things around to get a nicer module structure for faster tests and whatnot?

**Review by ywelsch — DISMISSED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: either is fine with me.

**Review by original-brownbear — DISMISSED** [review]
LGTM2, the other discussion can be had elsewhere IMO :)

**Comment by DaveCTurner** [comment]
I've just noticed a consequence of moving away from `delegate_type` to options on a standard type, combined with our general lack of validation that all the provided settings make sense: if you do not have X-pack installed and you try to define one of these magic repositories then you will silently get a normal repository. I think that is a blocker here.

**Comment by ywelsch** [comment]
I wonder if things are truly worse than with what we have today. Except for the type, we do not check if extra settings on a repository make any sense. I think we should move towards validating these settings, but at the same time also don't think this should be a blocker for this PR.

**Comment by original-brownbear** [comment]
> I wonder if things are truly worse than with what we have today.

We could create some weird situations where you create a repo with x-pack accidentally disabled. Then fix that, add x-pack and now the repo does something different. Especially for `encrypted` this could have some nasty side effects I guess? (though obviously we should account for reusing a repo path this way by other means as well ... maybe it's not so bad and just annoying)

I'm also fine with not blocking this I think if you don't think the above is annoying enough :)

**Comment by DaveCTurner** [comment]
Yes I think this would make things worse than they are today. We don't plan to guarantee any kind of sensible behaviour if you have multiple nodes writing to a repository that do not all agree on whether it's encrypted/compressed/source-only, nor do we guarantee that nodes have a consistent set of plugins installed. Today, the validation of the `type` field ensures that all nodes have the right plugins, but this proposal will removing that check.

I think it would be enough to define a single new "composite" repository type and stick with the `delegate_type` field for now. I'm struggling a bit to come up with a good name from the user's point of view. `composite` is not really helpful. `advanced`? `extended`? Suggestions welcome.

**Comment by original-brownbear** [comment]
> I think it would be enough to define a single new "composite" repository type and stick with the delegate_type field for now.

👍 I like this. It definitely removes a lot of risks.

> composite is not really helpful.

"delegating" ? Idk, composite is fine by me as well though. The name doesn't help/matter much does it? You'll have to track down the documentation of how `delegate_type` works anyway IMO.

**Comment by DaveCTurner** [comment]
I'm not going to pursue this PR, at least not for now. It's not needed for searchable snapshots any more.

**Pull request closed without merging**